packages/component-annotate-plugin/src/index.ts

```diff
@@ -0,0 +1,510 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+/**
+ * The following code is based on the FullStory Babel plugin, but has been modified to work
+ * with Sentry products
+ */
+
+import type * as Babel from "@babel/core";
+import type { PluginObj, PluginPass } from "@babel/core";
+
+import { DEFAULT_IGNORED_ELEMENTS, KNOWN_INCOMPATIBLE_PLUGINS } from "./constants";
+
+const webComponentName = "data-sentry-component";
+const webElementName = "data-sentry-element";
+const webSourceFileName = "data-sentry-source-file";
+
+const nativeComponentName = "dataSentryComponent";
+const nativeElementName = "dataSentryElement";
+const nativeSourceFileName = "dataSentrySourceFile";
+
+interface AnnotationOpts {
+  native?: boolean;
+  "annotate-fragments"?: boolean;
+  ignoreComponents?: IgnoredComponent[];
+}
+
+interface AnnotationPluginPass extends PluginPass {
+  opts: AnnotationOpts;
+}
+
+type IgnoredComponent = [file: string, component: string, element: string];
+
+type AnnotationPlugin = PluginObj<AnnotationPluginPass>;
+
+export function componentNameAnnotatePlugin({ types: t }: typeof Babel): AnnotationPlugin {
+  return {
+    visitor: {
+      FunctionDeclaration(path, state) {
+        if (!path.node.id || !path.node.id.name) {
+          return;
+        }
+        if (isKnownIncompatiblePluginFromState(state)) {
+          return;
+        }
+
+        functionBodyPushAttributes(
+          state.opts["annotate-fragments"] === true,
+          t,
+          path,
+          path.node.id.name,
+          sourceFileNameFromState(state),
+          attributeNamesFromState(state),
+          state.opts.ignoreComponents ?? []
+        );
+      },
+      ArrowFunctionExpression(path, state) {
+        const parent = path.parent; // We're expecting a `VariableDeclarator` like `const MyComponent =`
+        if (
+          !parent ||
+          !("id" in parent) ||
+          !parent.id ||
+          !("name" in parent.id) ||
+          !parent.id.name
+        ) {
+          return;
+        }
+
+        if (isKnownIncompatiblePluginFromState(state)) return;
```

Comment:
```suggestion
        if (isKnownIncompatiblePluginFromState(state)) {
          return;
        }
```

ditto with other `return` cases.